examples/Cassie/cassie_rbt_state_estimator.cc

```diff
@@ -816,32 +848,164 @@ void CassieRbtStateEstimator::contactEstimation(
+        position_index_map_.at("knee_joint_left"));
+    double right_knee_spring = output.GetPositionAtIndex(
+        position_index_map_.at("knee_joint_right"));
+    double left_heel_spring = output.GetPositionAtIndex(
+        position_index_map_.at("ankle_spring_joint_left"));
+    double right_heel_spring = output.GetPositionAtIndex(
+        position_index_map_.at("ankle_spring_joint_right"));
+    if (left_knee_spring < knee_spring_threshold_ekf_ &&
+          left_heel_spring < heel_spring_threshold_ekf_) {
+      *left_contact = 1;
+    }
+    if (right_knee_spring < knee_spring_threshold_ekf_ &&
+          right_heel_spring < heel_spring_threshold_ekf_) {
+      *right_contact = 1;
+    }
+
+    if (print_info_to_terminal_) {
+      cout << "left/right knee spring, threshold = " <<
+          left_knee_spring << ", " << right_knee_spring << ", " <<
+          knee_spring_threshold_ekf_ << endl;
+      cout << "left/right heel spring, threshold = " <<
+          left_heel_spring << ", " << right_heel_spring << ", " <<
+          heel_spring_threshold_ekf_ << endl;
+      cout << "left/right contacts = " <<
+          *left_contact << ", " << *right_contact << endl;
+    }
+  }
+}
+
+
+/// EstimateContactForController(). Less conservative.
+///
+/// Input:
+///  - OutputVector `output` containing the positions, velocities and
+///    actuator torques of the robot
+/// Ouput: left contact `left_contact` and right contact `right_contact` that
+///  indicate if the corresponding foot is in contact with the ground
+///
+/// Assumptions:
+///  1. the swing leg doesn't stop during single support
+///  2. flight mode is not present during the gait
+///
+/// Algorithm:
+///  The contact is estimated based on
+///   1. compression in the spring
+///   2. three optimizations one for each double support, left support and
+///   right support
+///  If the compression in the left (right) heel/ankle spring is more than the
+///  set threshold, the left (right) foot is estimated to be in contact with
+///  the ground.
+///  Additionally, the optimal cost from the optimization is used to augument
+///  the estimation from the spring.
+///  During impact, both the legs have some non-zero acceleration and hence the
+///  optimal costs will all be high. This case is assumed to be double stance.
+void CassieRbtStateEstimator::EstimateContactForController(
+    const OutputVector<double>& output,
+    int* left_contact, int* right_contact) const {
   // Estimate contact based on optimization results
   // The vector optimal_cost has double support, left support and right support
   // costs in order. The corresponding indices are 0, 1, 2.
   // Here we get the index of min of left and right support costs.
-  auto min_it = std::min_element(std::next(optimal_cost.begin(), 1),
-      optimal_cost.end());
-  int min_index = std::distance(optimal_cost.begin(), min_it);
+  auto min_it = std::min_element(std::next(optimal_cost->begin(), 1),
```

Comment:
This seems like overkill to get whether the left support or right support has the lower cost. An alternative could be `int min_index = optimal_cost->at(2) > optimal_cost->at(1) + 1;`. Although neither option is very readable.

Reply:
Good catch. I think this might be a bug. The intension was to get the min_index among the three costs.